### DISCUSSION TITLE
Support for custom internal webserver middlewares

### DISCUSSION THREAD

**Discussion by pfk84** [clue/framework-x#190] [Ideas]
In regards to the built-in limit of the ReactPHP http server with its limit of 64KB for the request body I'd love to have an option to increase/remove this limit without having to patch the Framework X source files, see also #69.

Right now there's nothing that can be configured in $app and the constructor just takes the request middlewares so to keep things simple I'd suggest to have the `run()` function accept middlewares as well that then are being passed to `runLoop()`.

This way one could add the already existing ReactPHP http server middlewares that change the default behaviours/limits (for example [RequestBodyBufferMiddleware](https://github.com/reactphp/http#requestbodybuffermiddleware)).

What do you think?

**Comment by pfk84** [comment]
Created a PR with a suggestion of how I'd do it: #193

**Comment by SimonFrings** [comment]
@pfk84 Seems like this will only work with the built-in webserver then. As I see it, this feature will eventually find its way into Framework x, so if we implement it your way as you suggested above, we would have to revert at some point. This would lead to BC-Breaks which we should avoid if we already know about them coming up. (also concerning https://github.com/clue/framework-x/pull/193)

My question now is: Isn't patching the Framework x files, for now, good enough? I can understand that it becomes a bit annoying and feels avoidable, but I think it's the best answer at the moment. I would also say we should let more people know about this, maybe with documenting this first as a possible alternative and then tackling the bigger topic at a later point. I'm intentionally saying "bigger topic" here because this involves quite a lot of work if we want to do this the right way (which should always be the way to do it!).

We're currently involved in planning the next major version for ReactPHP which will also benefit Framework X, so to be completely honest with you, it will take some time before we can look further into this.

**Reply by pfk84** [reply]
Well, yes, that's the whole point that it only applies to the built-in webserver... :)

As X's internal webserver is not something special to X but simply ReactPHP http server I think that it definitely makes sense to have more (direct) control over it, in any case.

I agree about the bigger topic but I have a feeling that most if not all work regarding upload/huge request body handling should be implemented in ReactPHP's http component anyway (probably as a cache-to-disk middleware), so even or especially then it should make sense have more control over it. Or am I completely off track? ;)

**Reply by SimonFrings** [reply]
> Or am I completely off track? ;)

No you're not, this is good input and brings this topic forward :+1:

>  agree about the bigger topic but I have a feeling that most if not all work regarding upload/huge request body handling should be implemented in ReactPHP's http component anyway (probably as a cache-to-disk middleware)...

Yes, I agree that it makes sense to set a custom limit in  [reactphp/http](https://github.com/reactphp/http), this is already possible as described in https://github.com/reactphp/http#httpserver:

> In order to override the above buffering defaults, you can configure the HttpServer explicitly. You can use the [LimitConcurrentRequestsMiddleware](https://github.com/reactphp/http#limitconcurrentrequestsmiddleware) and [RequestBodyBufferMiddleware](https://github.com/reactphp/http#requestbodybuffermiddleware) (see below) to explicitly configure the total number of requests that can be handled at once like this:

 As I also wrote in https://github.com/clue/framework-x/pull/196, Framework X should handle this for you in the future to avoid configuring everything all by yourself from scratch.

This was also brought up multiple times in the reactphp/http component, e.g. https://github.com/reactphp/http/issues/412#issuecomment-886260083.
